### PR TITLE
[quant] Remove qconfig_dict from API

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23465 [quant] Remove qconfig_dict from API**

Summary:
We decided not to allow user to use qconfig_dict to do quantization
since that API is not robust.

Test Plan:
```
python test/test_quantization.py
```

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

Differential Revision: [D16611504](https://our.internmc.facebook.com/intern/diff/D16611504)

### CODE REVIEW

File: torch/quantization/quantize.py [hunk 1]
```diff
@@ -125,28 +125,24 @@ def add_quant_dequant(module):
         module._modules[name] = add_quant_dequant(child)
     return module
 
-def prepare(model, qconfig_dict=None):
-    r"""Prepares the model for calibration or training given a qconfig_dict.
+def prepare(model):
+    r"""Prepares the model for calibration or training.
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: you probably want to expand comments on qconfig arguments being set on modules. Same comment for other functions below

**Review by dzhulgakov — COMMENTED** [review]
It basically sets the precedent of using `.qconfig` as a main API to provide settings on modules (before it was more of an implementation detail). It seems ok-ish, but want to check with @soumith @gchanan for the precedent of throwing random attributes on other python classes. I wonder whether we should add qconfig=None to the nn.Module base class to be explicit.

The user code looks like:
```
m = MyModel()
m.fc1.qconfig = default_qconfig
m = quantize(m)
```

File: test/common_quantization.py [hunk 2]
```diff
@@ -146,6 +165,99 @@ def forward(self, x):
...
+        self.fc3 = torch.nn.Linear(5, 5).to(dtype=torch.float)
+        self.fc3.qconfig = default_qconfig
+
+    def forward(self, x):
+        x = self.sub1(x)
+        x = self.sub2(x)
+        x = self.fc3(x)
+        return x
+
+class CustomeQConfigModel(torch.nn.Module):
```
Comment (raghuramank100) [inline test/common_quantization.py]: NIT: Typo: CustomQConfig

File: test/common_quantization.py [hunk 3]
```diff
@@ -146,6 +165,99 @@ def forward(self, x):
...
+class AnnotatedCustomConfigNestedModel(torch.nn.Module):
+    def __init__(self):
+        super(AnnotatedCustomConfigNestedModel, self).__init__()
+        self.sub1 = LinearReluModel()
+        self.sub2 = TwoLayerLinearModel()
+        self.fc3 = QuantWrapper(torch.nn.Linear(5, 5).to(dtype=torch.float))
+        self.fc3.qconfig = default_qconfig
+        self.sub2.qconfig = default_qconfig
+
+        custum_options = {
```
Comment (raghuramank100) [inline test/common_quantization.py]: NIT: custom_options

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please request feedback from @soumith and @gchanan on having qconfig as a field for all modules in nn.

File: test/common_quantization.py [hunk 4]
```diff
@@ -146,6 +165,99 @@ def forward(self, x):
...
+        x = self.sub2(x)
+        x = self.fc3(x)
+        return x
+
+class CustomeQConfigModel(torch.nn.Module):
+    def __init__(self):
+        super(CustomQConfigModel, self).__init__()
+        self.sub1 = LinearReluModel()
+        self.sub2 = QuantWrapper(TwoLayerLinearModel())
+        custum_options = {
```
Comment (resistor) [inline test/common_quantization.py]: s/custum/custom/

File: test/common_quantization.py [hunk 5]
```diff
@@ -146,6 +165,99 @@ def forward(self, x):
...
+            'dtype': torch.quint8,
+            'qscheme': torch.per_tensor_affine
+        }
+        custom_qconfig = QConfig(weight=default_weight_observer(),
+                                 activation=default_observer(**custum_options))
```
Comment (resistor) [inline test/common_quantization.py]: s/custum/custom/

File: test/common_quantization.py [hunk 6]
```diff
@@ -156,9 +268,9 @@ def __init__(self):
     def forward(self, x):
         return self.relu(self.fc2(self.relu(self.fc1(x))))
 
-class WrappedModel(torch.nn.Module):
+class SkipQuantModel(torch.nn.Module):
```
Comment (resistor) [inline test/common_quantization.py]: If feel that if the name can be changed this significantly, then maybe it should have a comment explaining what it is.

File: test/common_quantization.py [hunk 7]
```diff
@@ -146,6 +165,99 @@ def forward(self, x):
...
+        self.fc3 = QuantWrapper(torch.nn.Linear(5, 5).to(dtype=torch.float))
+        self.fc3.qconfig = default_qconfig
+        self.sub2.qconfig = default_qconfig
+
+        custum_options = {
+            'dtype': torch.quint8,
+            'qscheme': torch.per_tensor_affine
+        }
+        custom_qconfig = QConfig(weight=default_weight_observer(),
+                                 activation=default_observer(**custum_options))
```
Comment (resistor) [inline test/common_quantization.py]: s/custum/custom/

Comment (jerryzh168) [reply test/common_quantization.py]: fixed

File: test/common_quantization.py (hunk 6 above)
Comment (jerryzh168) [reply test/common_quantization.py]: sure, it's just we changed API and previously only this model is using QuantWrapper, and now every model uses it, so need to pick something else to distinguish this model with others.

**Comment by jerryzh168** [comment]
@dzhulgakov btw we rely on qconfig being not present right now. if module.qconfig=None is already set, we assume user don't want to quantize this module, see SkipQuantModel example in common_quantization.py

**Comment by gchanan** [comment]
at least for Modules, I don't think we have the precedence of adding random attributes to change the behavior -- those are usually done through hooks/apply methods/eval/train etc.

**Comment by jerryzh168** [comment]
we don't need them to be default attributes in nn.Module, this is used for eager mode quantization only

**Comment by dzhulgakov** [comment]
Unfortunately I don’t see strictly better option than setting per module qconfig setting now. Other options are not much cleaner.

The only suggestion might be to have a top level function ‘mark_for_quantization(module, qconfig)’ - at least it gives us ability to change implementation later if needed

Similarly - the question is whether we should do recursive propagation automatically or just ask user to do apply() explicitly (the latter allows naturally filter based on module types or any other conditions too)

**Comment by jerryzh168** [comment]
@dzhulgakov you mean
```python
def mark_for_quantization(module, qconfig):
     module.qconfig = qconfig
```
?

**Comment by jerryzh168** [comment]
I think we can have 
```
def mark_for_quantization_helper(qconfig):
     def mark_qconfig(module):
           module.qconfig = qconfig
     return mark_qocnfig

def mark_for_quantization(module, qconfig):
     module.apply(mark_for_quantization(qconfig))
```
and remove propogate qconfig?

**Comment by jerryzh168** [comment]
actually, if we do this, the actual qconfig of a submodule will be determined by the order they do mark_quantization,
```python
mark_quantization(sub, qconfig1)
mark_quantization(sub.fc1, qconfig2)
```
and 
```python
mark_quantization(sub.fc1, qconfig2)
mark_quantization(sub, qconfig1)
```
will result in different `qconfig` for `sub.fc1`, I guess it is still valuable to do `propagate_qconfig`?

**Comment by dzhulgakov** [comment]
I don't know, there are no obviously better way. Several ways that I can think of:

1) today's version
```
sub.qconfig = c1
sub.fc1.qconfig = c2
# doing propagation
```

2) just wrap it into a function
```
def annotate(module, qconfig):
     module.qconfig = qconfig

# order can be arbitrary
quantization.annotate(sub, c1)
quantization.annotate(sub.fc1, c2)
# doing propagation
```

3) use recursive application explicitly (it's more verbose but is kind of obvious
```
# order matters
sub.apply(lambda m: quantization.annotate(m, c1))
quantization.annotate(sub.fc1, c2)
# no propagation

# alternatively, it's easy to write any filters, e.g.:
for m in sub.modules():
  quantization.annotate(m, c1 if isinstance(m, nn.Linear) else c2)

# we might also add `overwrite` flag to quantization.annotate and without it error out if annotation is already there
```

I'm  not sure how to break a tie. We can merge current option (1) first and see feedback. Or maybe @gchanan @soumith have more insights.

**Comment by jerryzh168** [comment]
right now we don't add qconfig if it is already there, this is required if we want the config of submodules overwrite the configs of parent.

option 1 and 2 doesn't have too much difference, I think it is fine as a eager mode API, we can allow more flexibility in graph mode.

**Review by raghuramank100 — APPROVED** [review]

**Review by z-a-f — APPROVED** [review]

**Pull request closed without merging**